
```{r "qc setup"}

# Summary files
stats_dir <- here::here(res_dir, "stats/")

fq_sum     <- str_c(stats_dir, proj, "_fastqc.tsv")
trim_sum   <- str_c(stats_dir, proj, "_cutadapt.tsv")
bwt_sum    <- str_c(stats_dir, proj, "_aligned.tsv")
dedup_sum  <- str_c(stats_dir, proj, "_", index_map, "_UMI_dedup.tsv")
counts_sum <- str_c(stats_dir, proj, "_featureCounts.tsv")
sub_sum    <- str_c(stats_dir, proj, "_results.tsv")


# Sample names and groups
qc_sams <- plot_grpss$SAMPLES %>% bind_rows()


# Create table of sample groups
# mutate(sample = str_remove(sample, str_c("-", grp)))
grp_key <- qc_sams %>%
  imap_dfr(~ crossing(grp = .y, sample = .x))

# QC theme
qc_theme <- theme(
  panel.border = element_blank(),
  axis.line    = element_line(color = "black", linewidth = ln_pt)
)

```

## FastQC

`FastQC` was used to assess the quality of each fastq file. A summary of the results is shown below.

```{r "fq summary", fig.width = 24, fig.height = 9}

# Some sample names are missing part of the fastq file name
# need to match samples with fastq files
fq_sum <- fq_sum %>%
  read_tsv(col_names = c("value", "metric", "sample"),show_col_types = FALSE) 

fq_sum <- fq_sum %>% 
  mutate(sample = sapply(sample, function(x) grp_key$sample[str_starts(x,grp_key$sample)][1])) %>%
  full_join(grp_key, by = c("sample"))

fq_sum <- fq_sum %>%
  mutate(
    value = fct_relevel(value, distinct(fq_sum,value)$value),
    grp   = fct_relevel(grp, unique(grp_key$grp))
  )

# Create tiles
fq_sum %>%
  ggplot(aes(sample, metric, fill = value)) +
  geom_tile(color = "white", size = 0.5) +
  facet_wrap(~ grp, scales = "free_x", nrow = 2) +
  scale_fill_manual(values = theme_colors[c(3, 4, 2)]) +
  theme_info +
  theme(
    legend.position = "top",
    legend.title = element_blank(),
    axis.title   = element_blank(),
    axis.text.x  = element_text(size = 8, angle = 45, hjust = 1, vjust = 1)
  ) +
  qc_theme



```

<br>

## Cutadapt

`Cutadapt` was used to trim adapters. Reads that were untrimmed or were too short were removed, metrics are shown below. The number of reads written is displayed on each bar.

```{r "cutadapt summary", fig.width = 16, fig.height = 10}

# Cutadapt stats
mets <- c(
  "Total_read_pairs_processed",
  "Pairs_written_(passing_filters)",
  "Pairs_that_were_too_short"
)

trim_sum <- trim_sum %>%
  read_tsv(col_names = c("sample", "metric", "value"),show_col_types = FALSE) %>%
  dplyr::filter(metric %in% mets) %>%
  separate(.,value,c("value","temp"),"\\(",extra="drop",fill="right",convert=T) %>% 
  dplyr::select(-temp) %>% 
  pivot_wider(names_from = "metric", values_from = "value") %>% 
  dplyr::mutate(Untrimmed = `Total_read_pairs_processed` - (`Pairs_that_were_too_short` + `Pairs_written_(passing_filters)`)) %>%
  dplyr::select(-`Total_read_pairs_processed`) %>%
  pivot_longer(cols = -sample, names_to = "metric")

# Create bar graphs
trim_sum %>%
  create_qc_bars(
    grp_df     = grp_key,
    grp_lvls   = names(qc_sams),
    sam_lvls   = NULL,
    met_lvls   = rev(c(mets[-1], "Untrimmed")),
    plot_cols  = theme_colors,
    lab_metric = "Pairs_written_(passing_filters)",
    n_rows     = 2
  ) +
  qc_theme

```


<br>

## Read alignment

`Bowtie2` was used to align reads, metrics are shown below. The number of aligned reads (aligned exactly 1 time + aligned >1 times) is displayed on each bar.

```{r "bwt summary", fig.width = 16, fig.height = 10}

# Bowtie2 stats
mets <- c(
  "aligned exactly 1 time",
  "aligned >1 times",
  "aligned 0 times"
)

bwt_sum <- bwt_sum %>%
  read_tsv(col_names = c("sample", "metric", "value"),show_col_types = FALSE) %>%
  dplyr::filter(metric %in% mets) %>%
  dplyr::mutate(value = as.numeric(value))

# Create bar graphs
bwt_sum %>%
  create_qc_bars(
    grp_df     = grp_key,
    grp_lvls   = names(qc_sams),
    sam_lvls   = NULL,
    met_lvls   = rev(mets),
    plot_cols  = theme_colors[c(2, 5, 4)],
    lab_metric = c("aligned exactly 1 time", "aligned >1 times"),
    n_rows     = 2
  ) +
  qc_theme

```


<br>

## PCR duplicates

Duplicate reads were removed using `umi-tools`, metrics are shown below. The number of reads written is displayed on each bar. 

```{r "dedup summary", fig.width = 16, fig.height = 10}

# dedup stats
mets <- c("Input Reads", "Number of reads out")

dup_sum <- dedup_sum %>%
  read_tsv(col_names = c("sample", "metric", "value"),show_col_types = FALSE) %>%
  dplyr::filter(metric %in% mets) %>%
  pivot_wider(names_from = "metric", values_from = "value") %>%
  dplyr::mutate(
    `Duplicated reads` = `Input Reads` - `Number of reads out`,
    sample=str_remove(sample, paste0("_",index_map,"_UMI"))
  ) %>%
  dplyr::select(-`Input Reads`) %>%
  pivot_longer(cols = -sample, names_to = "metric")

# Create bar graphs
dup_sum %>%
  create_qc_bars(
    grp_df     = grp_key,
    grp_lvls   = names(qc_sams),
    sam_lvls   = NULL,
    met_lvls   = rev(c(mets[-1], "Duplicated reads")),
    plot_cols  = c(theme_colors[1], "#C4C4C4"),
    lab_metric = "Number of reads out",
    n_rows     = 2
  ) +
  qc_theme

```

<br>

## Read assignment

`featureCounts` was used to calculate gene counts. A summary of the contribution of different RNAs to each library is shown below. The number of reads aligning within protein coding genes is displayed on each bar.

```{r "featurecounts", fig.width = 16, fig.height = 10}

# RNAs to summarize
rnas <- c(
  "protein_coding", "miRNA",
  "unassigned",     "rRNA",
  "snoRNA",         "snRNA"
)

# Calculate unassigned counts
un_counts <- str_c(counts_sum, ".summary") %>%
  read_tsv(.,show_col_types = FALSE) %>%
  pivot_longer(cols = ends_with(".bam"), names_to = "sample") %>%
  trim_path(paste0("_",index_map,"\\.bam$")) %>%
  
  dplyr::filter(grepl("^Unassigned_", Status)) %>%
  dplyr::mutate(metric = "unassigned") %>%
  
  group_by(sample, metric) %>%
  summarize(value = sum(value), .groups = "drop")

# featureCounts stats

counts_sum <- counts_sum %>%
  read_tsv(skip = 1,show_col_types = FALSE) %>%
  pivot_longer(cols = ends_with(".bam"), names_to = "sample") %>%
  trim_path(paste0("_",index_map,"\\.bam$")) %>%
  mutate(name = str_c(Geneid, gene_name, sep = "|")) %>%
  dplyr::select(
    name, Length, sample,
    metric = gene_biotype,
    value
  ) %>%
  
  bind_rows(un_counts) %>%  # add unassigned reads
  
  mutate(
    metric = str_replace(metric, "Mt_rRNA", "rRNA"),
    metric = ifelse(!metric %in% rnas, "other", metric)
  ) %>%
  
  group_by(sample, metric) %>%
  summarize(value = sum(value), .groups = "drop") 


# Create bar graphs
counts_sum %>%
  create_qc_bars(
    grp_df     = grp_key,
    grp_lvls   = names(qc_sams),
    sam_lvls   = NULL,
    met_lvls   = rev(c(rnas, "other")),
    plot_cols  = theme_colors,
    lab_metric = "protein_coding",
    n_rows     = 2
  ) +
  qc_theme

```

<br>

## Subsampling

Unique reads were filtered of reads aligned to snoRNA genes. Libraries were then subsampled with samtools so that downstream analysis is performed using an equal number of unique reads for sample groups. The number of sampled reads is displayed on each bar.

```{r "Sub summary", fig.width = 16, fig.height = 10}

# Subsampling stats
mets <- c("Sampled reads", "Reads removed")

sub_sum <- sub_sum %>% read_tsv(.,show_col_types = FALSE) %>% 
  select(sample,starts_with("aligned",ignore.case = F),starts_with("sub")) 

if(any(names(sub_sum) %in% "sub_group")){
  sub_sum <- sub_sum %>%
    dplyr::rename(grp = sub_group,
                  `Filtered reads` = starts_with("aligned",ignore.case = F), 
                  `Sampled reads` = starts_with("subsampled_")) %>% 
    separate(.,`Sampled reads`,c("Sampled reads","temp"),"\\(",extra="drop",fill="right",convert=T) %>% 
    dplyr::select(-temp) 
} else {
  sub_sum <- sub_sum %>% 
    dplyr::rename(`Filtered reads` = starts_with("aligned",ignore.case = F)) %>% 
    dplyr::mutate(`Sampled reads` = `Filtered reads`) %>%
  full_join(grp_key, by = c("sample"))
}

sub_sum <- sub_sum %>% 
  dplyr::mutate(`Reads removed` = `Filtered reads` - `Sampled reads`) %>%
  pivot_longer(cols = -c(grp, sample), names_to = "metric") %>%
  dplyr::filter(metric %in% mets)



# Create bar graphs
sub_sum %>%
  create_qc_bars(
    grp_lvls   = names(qc_sams),
    sam_lvls   = NULL,
    met_lvls   = rev(mets),
    plot_cols  = theme_colors,
    lab_metric = "Sampled reads",
    n_rows     = 2
  ) +
  qc_theme

```

---

<br>

<br>
