# ====== Rules for aligning reads =================================


# Align trimmed reads with star
rule star_twopass:
    input:
        R1 = PROJ + "/{sample}_R1.fastq.gz",
        R2 = PROJ + "/{sample}_R2.fastq.gz"
    output:
        bam   = temp(PROJ + "/{sample}_" + INDEX_MAP + ".bam"),
        bai   = temp(PROJ + "/{sample}_" + INDEX_MAP + ".bam.bai"),
        stats = PROJ + "/bams/{sample}_Log.out"
    params:
        job_name = "{sample}_star_twopass",
        gtf      = GTF,
        bam_out  = PROJ + "/bams/{sample}_Aligned.sortedByCoord.out.bam",
        args     = CMD_PARAMS["star"],
        out      = PROJ + "/bams/{sample}_"
    resources:
        memory = lambda wildcards, input: memory_estimator([input.R1, input.R2], 3, 30)
    log:
        out = PROJ + "/logs/{sample}_star_twopass.out",
        err = PROJ + "/logs/{sample}_star_twopass.err"
    message:
        "Running STAR alignment with two-pass mode"
    threads: 
        16
    shell:
        """
        STAR \
          --genomeDir {INDEX_PATH} \
          --runThreadN {threads} \
          --readFilesIn {input.R1} {input.R2} \
          --outSAMtype BAM SortedByCoordinate \
          --limitBAMsortRAM {resources.memory}000000000 \
          --readFilesCommand gunzip -c \
          --outFileNamePrefix {params.out} \
          --sjdbGTFfile {params.gtf} \
          {params.args} \
          --twopassMode Basic
          
        mv {params.bam_out} {output.bam}

        samtools index -@ {threads} {output.bam}
        """
        

# ====== Rules for aligning reads summmary =================================


# Align trimmed reads with star summmary
rule star_2pass_summmary:
    input:
        expand(
           PROJ  + "/bams/{sample}_Log.out",
            sample = SAMS_UNIQ
        )
    output:
        PROJ + "/stats/" + PROJ + "_aligned.tsv"
    params:
        job_name = "star_2pass_summmary"
    resources:
        memory   = 1
    log:
        out = PROJ + "/logs/star_summmary.out",
        err = PROJ + "/logs/star_summmary.err"
    message:
        "summmary star alignments "
    threads: 
        4
    threads:
        1
    run:
        with open(output[0], "w") as out:
            for file in input:
                name = os.path.basename(file)
                name = re.sub("_2pass_Log.final.out", "", name)

                for line in open(file, "r"):
                    match = re.search(r"Uniquely mapped reads % \|\s+(.*)%", line)

                    if match:
                        num = match.group(1) + "%"
                        met = "overall alignment rate"

                        out.write("%s\t%s\t%s\n" % (name, met, num))
                        
                        


                    